### PR TITLE
Add lite option

### PR DESCRIPTION
**Pull request by uilianries**
Hi!

For this PR I added lite option, thanks to @elizagamedev 

Also, I've added the option `protoc` that creates **protoc** and **libprotoc**, but libprotoc is not listed in cpp_info.libs

### CODE REVIEW

**Comment by uilianries** [comment]
@jgsogo @SSE4 I think this PR is good to be merged, but I would like your thoughts before

There are two new options: `lite` and `protoc`.

**lite** is used to switch between `libprotobuf-lite` and `libprotobuf`. When **lite** is True, **ONLY** `libprotobuf-lite` is present in `self.cpp_info.libs`. Otherwise, only `libprotobuf` will be listed there. As `protoc` requires `libprotobuf`, it's necessary build both lite and regular versions.

**protoc** option is used to disable `protoc` and `libprotoc` build. When this option is False, only `libprotobuf` or `libprotobuf-lite` are built and distributed. Even when **protoc** is True, `libprotoc` is **NOT** listed in `self.cpp_info.libs`.

Both `protobuf_generate_cpp` (protobuf_MODULE_COMPATIBLE for legacy mode) and `protobuf_generate` are working without black magics.

File: conanfile.py [hunk 1]
```diff
@@ -69,56 +81,66 @@ def build(self):
...
-            libprotobuf = "libprotobufd.%s.dylib" % self.version if self.settings.build_type == "Debug"\
-                else "libprotobuf.%s.dylib" % self.version
+            suffix = "d" if self.settings.build_type == "Debug" else ""
+            libprotoc = "libprotoc%s.%s.dylib" % (suffix, self.version)
+            libprotobuf = "libprotobuf%s.%s.dylib" % (suffix, self.version)
             for lib in [libprotoc, libprotobuf]:
                 command = "install_name_tool -change %s @executable_path/../lib/%s %s" % (lib, lib, protoc)
                 self.run(command)
             libprotoc = os.path.join(self.package_folder, "lib", libprotoc)
             command = "install_name_tool -change %s @loader_path/%s %s" % (libprotobuf, libprotobuf, libprotoc)
```
Comment (jgsogo) [inline conanfile.py]: What happens if the user has something already installed? Are we overriding it system-wide?

File: conanfile.py [hunk 2]
```diff
@@ -41,7 +49,8 @@ def configure(self):
     def requirements(self):
         if self.options.with_zlib:
             self.requires("zlib/1.2.11@conan/stable")
-        self.requires("protoc_installer/3.6.1@bincrafters/stable")
+        if self.options.protoc:
+            self.requires("protoc_installer/3.6.1@bincrafters/stable")
```
[the remark this answers is not in the document]
Comment (jgsogo) [reply conanfile.py]: Totally agree, but there are many other packages that create several libraries and put them all in `cpp_info.libs`... it is an issue if you use `${CONAN_LIBS}` as it is like _gotta catch em all_, but you can always use the TARGETS approach and select which libraries to link.

It's a problem, but we are not creating it here.

File: conanfile.py (hunk 1 above)
Comment (jgsogo) [reply conanfile.py]: the append-to-path approach should we enough, or there is something else related to the CMake macros?

File: conanfile.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (uilianries) [reply conanfile.py]: done!

**Comment by uilianries** [comment]
@jgsogo Now only there is `lite` option.

as `protoc_installer` is no longer required, all black magic over cmake file is gone.

File: test_package/CMakeLists.txt [hunk 3]
```diff
@@ -6,6 +6,9 @@ conan_basic_setup()
 
 set(CMAKE_VERBOSE_MAKEFILE ON)
 
+option(ENABLE_PROTOC "Run protoc to generate sources" ON)
```
Comment (jgsogo) [inline test_package/CMakeLists.txt]: We don't have `protoc` here, it is in `protoc_installer`... so here we have to make the usual test: that the libraries have been compiled, that we can link against them, and run some code (maybe even create a dynamic protobuf).

File: protobuf.patch [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
++  include(libprotobuf-lite.cmake)
++endif (protobuf_BUILD_PROTOBUF_LITE)
++
++if (protobuf_BUILD_PROTOC_BINARIES OR NOT protobuf_BUILD_PROTOBUF_LITE)
++  include(libprotobuf.cmake)
++endif (protobuf_BUILD_PROTOC_BINARIES OR NOT protobuf_BUILD_PROTOBUF_LITE)
++
+ if (protobuf_BUILD_PROTOC_BINARIES)
+   include(libprotoc.cmake)
+   include(protoc.cmake)
```
[the remark this answers is not in the document]
Comment (uilianries) [reply protobuf.patch]: understood, I'll remove all protoc occurrences

File: protobuf.patch [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
++  list(APPEND _protobuf_libraries libprotobuf-lite)
++endif (protobuf_BUILD_PROTOBUF_LITE)
++
++if (protobuf_BUILD_PROTOC_BINARIES OR NOT protobuf_BUILD_PROTOBUF_LITE)
++  list(APPEND _protobuf_libraries libprotobuf)
++endif (if (protobuf_BUILD_PROTOC_BINARIES OR NOT protobuf_BUILD_PROTOBUF_LITE))
++
+ if (protobuf_BUILD_PROTOC_BINARIES)
+     list(APPEND _protobuf_libraries libprotoc)
+ endif (protobuf_BUILD_PROTOC_BINARIES)
```
[the remark this answers is not in the document]
Comment (uilianries) [reply protobuf.patch]: yeah ditto

File: test_package/CMakeLists.txt (hunk 3 above)
Comment (uilianries) [reply test_package/CMakeLists.txt]: sure!

**Comment by uilianries** [comment]
> Why don't we have in this repo the recipe for protoc_installer too? We can have two packages in one repo, thats not a problem. Why am I suggesting this? Because then, we will have full protobuf or nothing, both packages or zero, we will run package tests for each recipe, BUT ALSO we can run an integration test using the package from protobuf and the executable from protoc_installer. It will be two steps more in CI, but I think it would be very interesting.

I would prefer the second choice. We could require `protoc_installer` in test_package. I think it will be a great integration test.

**Comment by jgsogo** [comment]
I don¡t know what was my first option 😆 

Two recipes, each one with its own tests, and a directory with integration tests (requiring the `protoc`). It will be a repository to take into account when implementing new stuff in Conan (like the cross-compiling model) 🥇

**Comment by uilianries** [comment]
Ops! I've pushed into bincrafters/testing instead of uilianries/testing

**Pull request merged**